File: facade.go
// Certainly! Below is a complex example of how the Facade design pattern can be implemented in Go. In this example, let's consider a complex subsystem related to a travel booking system that includes multiple components such as flight booking, hotel reservation, and car rental. The TravelFacade will act as a simplified interface to these subsystems.

// Copy
package main

import "fmt"

// FlightBookingSubsystem
type FlightBooking struct {
}

func (fb *FlightBooking) bookFlight(destination string) {
	fmt.Printf("Booking a flight to %s\n", destination)
}

// HotelReservationSubsystem
type HotelReservation struct {
}

func (hr *HotelReservation) reserveHotel(city string, nights int) {
	fmt.Printf("Reserving a hotel in %s for %d nights\n", city, nights)
}

// CarRentalSubsystem
type CarRental struct {
}

func (cr *CarRental) rentCar(city string, days int) {
	fmt.Printf("Renting a car in %s for %d days\n", city, days)
}

// TravelFacade
type TravelFacade struct {
	flightBooking     *FlightBooking
	hotelReservation  *HotelReservation
	carRental         *CarRental
}

func NewTravelFacade() *TravelFacade {
	return &TravelFacade{
		flightBooking:    &FlightBooking{},
		hotelReservation: &HotelReservation{},
		carRental:        &CarRental{},
	}
}

func (tf *TravelFacade) PlanTrip(destination string, nights int, days int) {
	tf.flightBooking.bookFlight(destination)
	tf.hotelReservation.reserveHotel(destination, nights)
	tf.carRental.rentCar(destination, days)
}

func main() {
	// Client code using the TravelFacade
	travelFacade := NewTravelFacade()

	// Plan a trip to Paris with 3 nights stay and 5 days car rental
	travelFacade.PlanTrip("Paris", 3, 5)

	// Plan a trip to New York with 2 nights stay and 4 days car rental
	travelFacade.PlanTrip("New York", 2, 4)
}
// In this example:

// FlightBooking, HotelReservation, and CarRental represent complex subsystems related to travel.
// TravelFacade is the facade class that provides a simplified interface to these subsystems.
// The PlanTrip method in the TravelFacade class coordinates the interactions with the subsystems to plan a trip.
// The client code can use the TravelFacade to plan trips without dealing with the intricacies of flight booking, hotel reservation, and car rental. The Facade pattern helps to simplify the usage of complex subsystems by providing a higher-level, unified interface.

File: facade.java
// Certainly! The Facade design pattern is used to provide a simplified interface to a set of interfaces in a subsystem, making it easier to use. Here's a complex example of how the Facade pattern can be used in Java:

// Let's consider a multimedia subsystem that consists of various components like AudioPlayer, VideoPlayer, and ImageLoader. The Facade class named MultimediaFacade will provide a simplified interface to these components.

// Copy
// Complex subsystems
class AudioPlayer {
    public void playAudio(String fileName) {
        System.out.println("Playing audio: " + fileName);
    }
}

class VideoPlayer {
    public void playVideo(String fileName) {
        System.out.println("Playing video: " + fileName);
    }
}

class ImageLoader {
    public void loadImage(String fileName) {
        System.out.println("Loading image: " + fileName);
    }
}

// Facade class providing a simplified interface
class MultimediaFacade {
    private AudioPlayer audioPlayer;
    private VideoPlayer videoPlayer;
    private ImageLoader imageLoader;

    public MultimediaFacade() {
        this.audioPlayer = new AudioPlayer();
        this.videoPlayer = new VideoPlayer();
        this.imageLoader = new ImageLoader();
    }

    public void playMedia(String fileName, MediaType type) {
        switch (type) {
            case AUDIO:
                audioPlayer.playAudio(fileName);
                break;
            case VIDEO:
                videoPlayer.playVideo(fileName);
                break;
            case IMAGE:
                imageLoader.loadImage(fileName);
                break;
            default:
                System.out.println("Unsupported media type");
        }
    }
}

// Enum to represent media types
enum MediaType {
    AUDIO, VIDEO, IMAGE
}

// Client code
public class FacadePatternExample {
    public static void main(String[] args) {
        MultimediaFacade multimediaFacade = new MultimediaFacade();

        // Playing audio using the facade
        multimediaFacade.playMedia("song.mp3", MediaType.AUDIO);

        // Playing video using the facade
        multimediaFacade.playMedia("movie.mp4", MediaType.VIDEO);

        // Loading image using the facade
        multimediaFacade.playMedia("picture.jpg", MediaType.IMAGE);
    }
}
In this example:

AudioPlayer, VideoPlayer, and ImageLoader are the complex subsystems representing audio playback, video playback, and image loading, respectively.
MultimediaFacade is the Facade class that provides a simplified interface to these subsystems.
The playMedia method in the MultimediaFacade class handles the complexity of choosing the appropriate subsystem based on the media type.
The client code can use the MultimediaFacade to play audio, video, and load images without worrying about the intricacies of each subsystem. The Facade pattern helps to simplify the usage of complex subsystems by providing a higher-level, unified interface.